src/tactic/refine_list.lean

```diff
@@ -0,0 +1,118 @@
+-- Copyright (c) 2019 Lucas Allen. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Lucas Allen
+
+-- TODO Documentation in tactics.md
+
+import tactic.library_search
+import data.mllist
+
+/-
+`refine_list` is an extension of `library_search` which lists some applicable lemmas and theorems.
+-/
+
+namespace tactic
+
+--list_refine uses some functions from library_search
+open tactic.library_search
+
+/-- Returns the current tactic_state. -/
+meta def get_state : tactic tactic_state :=
+λ s, result.success s s
+
+/-- Sets the tactic_state. -/
+meta def set_state (s' : tactic_state) : tactic tactic_state :=
+λ s, result.success s s'
+
+/-- Runs a tactic, returning the result and the new tactic_state, but then reverts the tactic_state. -/
+-- See also `lock_tactic_state`, which is similar but does not return the tactic_state.
+meta def run_and_save_state {α : Type} (t : tactic α) : tactic (α × tactic_state)
+| s := match t s with
+       | result.success a s' := result.success (a, s') s
+       | result.exception msg pos s' := result.exception msg pos s
+end
+
+/--This function prints either the `exact` or `refine` tactics with the corresponding
+lemma/theorem with inputs for a specific tactic_state-/
+meta def message (l : decl_data × tactic_state) (g : expr) : tactic string :=
+do s ← get_state,
+   set_state l.2,
+   r ← tactic_statement g,
+   set_state s,
+   return r
+
+/--Runs through the list of tactic_states and prints an `exact` or `refine` message for each one-/
+meta def print_messages (g : expr) (silent : bool) : list (decl_data × tactic_state) → tactic (list string)
+| []      := return []
+| (l::ls) := do r ← message l g,
+                if ¬ silent then trace r else skip,
+                rs ← print_messages ls,
+                return (r :: rs)
+
+/-- Returns a monadic lazy list of declaration data -/
+meta def get_mldefs (defs : list decl_data) : mllist tactic decl_data :=
+mllist.of_list defs
+
+
+declare_trace silence_refine_list -- Turn off `exact ...` trace message
+declare_trace refine_list         -- Trace a list of all relevant lemmas
+
+/-- The main list_refine tactic, this is very similar to the main library_search function -/
```

Comment:
Remember that docstrings are most helpful when they show a message in a tooltip that is meaningful outside the context of this file.
Of course this is not the interactive tactic. I see that you have a very nice docstring over there. It's not so important, but I think this docstring could be improved by giving a very short (1 or 2 lines) summary of what this tactic does.
Finally, the docstring says `list_refine` instead of `refine_list`.

Reply:
Is this OK?

/-- The main refine_list tactic, this is very similar to the main library_search function. It returns a list of strings consisting of possible applications of the refine tactic. The length of the list is no longer than num.-/

Reply:
Yeah, looks good to me.